src/shape.rs

```diff
@@ -348,3 +376,356 @@ mod test
         assert_eq!(None, r1.intersect(&r2));
     }
 }
+
+///////////////////////////////////
+
+/// A struct representing an axis-aligned rounded rectangle. Two points and a float are
+/// stored: the top left vertex, the bottom right vertex and the radius of the
+/// rounded corners.
+///
+/// Alias for a rectangle with u32 coordinates.
+pub type URoundRect = RoundedRectangle<u32>;
+
+/// A struct representing an axis-aligned rounded rectangle. Two points and a float are
+/// stored: the top left vertex, the bottom right vertex and the radius of the
+/// rounded corners.
+///
+/// Alias for a rectangle with i32 coordinates.
+pub type IRoundRect = RoundedRectangle<i32>;
+
+/// A struct representing an axis-aligned rounded rectangle. Two points and a float are
+/// stored: the top left vertex, the bottom right vertex and the radius of the
+/// rounded corners.
+///
+/// Alias for a rectangle with f32 coordinates.
+pub type RoundRect = RoundedRectangle<f32>;
+
+/// A struct representing an axis-aligned rounded rectangle. Two points and a float are
+/// stored: the top left vertex, the bottom right vertex and the radius of the
+/// rounded corners.
+#[derive(Debug, PartialEq, Eq, Clone)]
+#[repr(C)]
+pub struct RoundedRectangle<T = f32>
+{
+    top_left: Vector2<T>,
+    bottom_right: Vector2<T>,
+    radius: T
+}
+
+impl<T> AsRef<RoundedRectangle<T>> for RoundedRectangle<T>
+{
+    fn as_ref(&self) -> &Self
+    {
+        self
+    }
+}
+
+impl<T> RoundedRectangle<T>
+{
+    /// Constructs a new `RoundedRectangle`. The top left vertex must be above
+    /// and to the left of the bottom right vertex. A negative radius will
+    /// be clamped to 0. A big radius (larger than half the width or height)
```

Comment:
This doesn't appear to be clamped inside `new()`

Reply:
hmm I wasn't sure where to check for a negative radius, I was thinking about the user setting a negative radius at runtime and breaking things up; I'll just add the check and add this clarification to the documentation